Injecting CustomTimeouts into TFBridge

stack72 opened the pull request.

pgavlin commented on pkg/tfbridge/provider.go:
```diff
@@ -928,3 +944,20 @@ func initializationError(id string, props *pbstruct.Struct, reasons []string) er
 	return rpcerror.WithDetails(rpcerror.New(codes.Unknown, reasons[0]), &detail)
 }
+
+func injectTimeoutValue(diff *terraform.InstanceDiff, timeout float64, timeoutKey string) *terraform.InstanceDiff {
```
Nit: `s/injectTimeoutValue/setTimeout`

stack72 replied:
Updated!

pgavlin commented on pkg/tfbridge/provider.go:
```diff
@@ -928,3 +944,20 @@ func initializationError(id string, props *pbstruct.Struct, reasons []string) er
+
+func setTimeout(diff *terraform.InstanceDiff, timeout float64, timeoutKey string) *terraform.InstanceDiff {
+	timeoutValue := int64(timeout * 1000000000) //this turns seconds to nanoseconds - TF wants it in this format
+	if diff.Meta == nil {
```
```suggestion
	if diff.Meta == nil {
		diff.Meta = map[string]interface{}{}
	}
	diff.Meta[schema.TimeoutKey] = map[string]interface{}{
		timeoutKey: timeoutValue,
	}
```

stack72 replied:
Much nicer :) Thanks!

stack72 commented:
ok, here we go.... 

Using this config:

```typescript
import * as aws from "@pulumi/aws";

const db = new aws.rds.Instance("testing", {
    allocatedStorage: 20,
    engine: "postgres",
    engineVersion: "11.4",
    instanceClass: aws.rds.InstanceTypes.R3_Large,
    password: "myTestPassw0rd!",
    username: "TestUser",
    skipFinalSnapshot: true,
}, {
    customTimeouts: {
        create: "1m",
    },
});

export const identifier = db.identifier;
```

before I updated AWS with the pulumi-terraform customTimeouts branch, I have the following:

```
MacBook-Pro :: Code/pulumi-testing/custom-timeout-testing % pulumi up --yes                                                                      255 ↵
warning: A new version of Pulumi is available. To upgrade from version '0.0.0' to '0.17.22', visit https://pulumi.io/install for manual instructions and release notes.
Previewing update (dev):

     Type                 Name                        Plan
     pulumi:pulumi:Stack  custom-timeout-testing-dev
 +   └─ aws:rds:Instance  testing                     create

Resources:
    + 1 to create
    1 unchanged

Updating (dev):

     Type                 Name                        Status
     pulumi:pulumi:Stack  custom-timeout-testing-dev
 +   └─ aws:rds:Instance  testing                     created

Outputs:
  + identifier: "testing56bd991"

Resources:
    + 1 created
    1 unchanged

Duration: 4m12s
```

As you can see, it ignored the 1minute timeout on create. After I updated AWS with the customTimeouts branch, I get the following:

```
MacBook-Pro :: Code/pulumi-testing/custom-timeout-testing % pulumi up --yes
warning: A new version of Pulumi is available. To upgrade from version '0.0.0' to '0.17.22', visit https://pulumi.io/install for manual instructions and release notes.
Previewing update (dev):

     Type                 Name                        Plan
 +   pulumi:pulumi:Stack  custom-timeout-testing-dev  create
 +   └─ aws:rds:Instance  testing                     create

Resources:
    + 2 to create

Updating (dev):

     Type                 Name                        Status                  Info
 +   pulumi:pulumi:Stack  custom-timeout-testing-dev  **creating failed**     1 error
 +   └─ aws:rds:Instance  testing                     **creating failed**     1 error

Diagnostics:
  pulumi:pulumi:Stack (custom-timeout-testing-dev):
    error: update failed

  aws:rds:Instance (testing):
    error: Plan apply failed: 1 error occurred:
    	* creating urn:pulumi:dev::custom-timeout-testing::aws:rds/instance:Instance::testing: timeout while waiting for state to become 'available, storage-optimization' (last state: 'creating', timeout: 1m0s)

Resources:
    + 1 created

Duration: 1m7s
```

As this has been given the green light, I am going to merge it now that we have proved it works!
